polars/polars-io/src/parquet/write.rs

```diff
@@ -90,63 +94,79 @@ where
                 // remaining is plain
                 _ => Encoding::Plain,
-            })
+            }
+        };
+
+        let encodings = (&schema.fields)
+            .par_iter()
+            .map(|f| transverse(&f.data_type, encoding_map))
             .collect::<Vec<_>>();
 
         let row_group_iter = rb_iter.filter_map(|batch| match batch.len() {
             0 => None,
             _ => {
-                let columns = batch
-                    .columns()
-                    .par_iter()
-                    .zip(parquet_schema.fields().par_iter())
-                    .zip(encodings.par_iter())
-                    .map(|((array, tp), encoding)| {
-                        let encoded_pages =
-                            array_to_pages(array.as_ref(), tp.clone(), options, *encoding)?;
-                        encoded_pages
-                            .map(|page| {
-                                compress(page?, vec![], options.compression).map_err(|x| x.into())
-                            })
-                            .collect::<ArrowResult<VecDeque<_>>>()
-                    })
-                    .collect::<ArrowResult<Vec<VecDeque<CompressedPage>>>>()
-                    .map(Some)
-                    .transpose();
-
-                columns.map(|columns| {
-                    columns.map(|columns| {
-                        let row_group = DynIter::new(
-                            columns
-                                .into_iter()
-                                .map(|column| Ok(DynStreamingIterator::new(Bla::new(column)))),
-                        );
-                        (row_group, batch.columns()[0].len())
-                    })
-                })
+                let row_group = create_serializer(
+                    batch,
+                    parquet_schema.fields().to_vec(),
+                    encodings.clone(),
+                    options,
+                );
+
+                Some(row_group)
             }
         });
 
         let mut writer = FileWriter::try_new(&mut self.writer, schema, options)?;
-        // write the headers
-        writer.start()?;
         for group in row_group_iter {
-            let (group, _len) = group?;
-            writer.write(group)?;
+            writer.write(group?)?;
         }
         let _ = writer.end(None)?;
 
         Ok(())
     }
 }
+
+fn create_serializer(
+    batch: Chunk<Arc<dyn Array>>,
+    fields: Vec<ParquetType>,
+    encodings: Vec<Vec<Encoding>>,
+    options: WriteOptions,
+) -> std::result::Result<RowGroupIter<'static, ArrowError>, ArrowError> {
+    let columns = batch
+        .columns()
+        .par_iter()
+        .zip(fields)
+        .zip(encodings.clone())
+        .flat_map(move |((array, type_), encoding)| {
+            let encoded_columns = array_to_columns(array, type_, options, encoding).unwrap();
+            encoded_columns
+                .into_iter()
```

Comment:
this could be `into_par_iter`, in which case even serializing nested columns get parallelized

Reply:
Starting a new `par_iter` in a parallel iterator, can lead to segfaults (actually just a SO, not so dramatic) I learned. 

If there are many parallel units of work and every unit throws a parallel iterator on the stack we might overflow. Because of work stealing there is no fifo guarantee.